EventListener/BreadcrumbListener.php

```diff
@@ -0,0 +1,402 @@
+    {
+        $this->addBreadcrumbsFromAnnotations($event, $annotations);
+    }
+
+    private function addBreadcrumbsFromAnnotations(
+        KernelEvent $event,
+        array $annotations,
+        ?string $routeName = null,
+        ?array $parameters = null
+    ) {
+        foreach ($annotations as $annotation) {
+            if (!$annotation instanceof Breadcrumb) {
+                continue;
+            }
+
+            if ($annotation->getParentRouteName()) {
+                $this->addBreadcrumbsForParent($annotation, $event);
+            }
+
+            $this->breadcrumbTrail->add(
+                $this->generateBreadcrumb(
+                    $event->getRequest(),
+                    $annotation,
+                    $routeName,
+                    $parameters
+                )
+            );
+        }
+    }
+
+    private function generateBreadcrumb(
+        Request $request,
+        Breadcrumb $breadcrumb,
+        ?string $route = null,
+        ?array $parameters = null
+    ): BreadcrumbValueObject {
+        $title = $breadcrumb->getTitle();
+        $routeParameters = $parameters ?? $breadcrumb->getRouteParameters();
+        $routeName = $route ?? $breadcrumb->getRouteName();
+        preg_match_all(
+            '#\{(?P<variable>\w+).?(?P<function>([\w\.])*):?(?P<parameters>(\w|,| )*)\}#',
+            $title,
+            $matches,
+            PREG_OFFSET_CAPTURE | PREG_SET_ORDER
+        );
+
+        // Convert the breadcrumb title to the correct value
+        foreach ($matches as $match) {
+            $varName = $match['variable'][0];
+            // Get the methods that need to be called on the route parameter
+            $functions = $match['function'][0] ? explode('.', $match['function'][0]) : [];
+            $parameters = $match['parameters'][0] ? explode(',', $match['parameters'][0]) : [];
+            $nbCalls = count($functions);
+
+            if ($request->attributes->has($varName)) {
+                $object = $request->attributes->get($varName);
+
+                // If no functions need to be called use the title as given
+                if (empty($functions)) {
+                    $objectValue = (string) $object;
+                    $title = str_replace($match[0][0], $objectValue, $title);
+
+                    continue;
+                }
+
+                // Chain the method calls so we can get the correct value
+                // Example book.author.name should call getAuthor().getName() on the given book object
+                foreach ($functions as $f => $function) {
+                    // While this is not the last function, call the chain
+                    if ($f < $nbCalls - 1) {
+                        $fullFunctionName = 'get' . $function;
+                        if (is_callable([$object, $fullFunctionName])) {
+                            $object = call_user_func([$object, $fullFunctionName]);
```

Comment:
Zit hier niet wat redundantie in?
```php
if (is_callable([$object, $fullFunctionName])) {
      $object = call_user_func([$object, $fullFunctionName]);
```
komt 6 keer voor, en 
```php
if (is_callable([$object, $fullFunctionName])) {
     $objectValue = call_user_func_array([$object, $fullFunctionName], $parameters);

     continue;
 }
```
ook 6 keer. Als die `$fullFunctionName = 'get' . $function;` het enige verschil is kan dit dan niet in een loop gebeuren met een array `["get", "has", "is"]` om de functienaam op te bouwen?

Reply:
Zou dit ook in een aparte private method stoppen onderaan de Listener en dan hier oproepen ofzo. Lijkt mij iets properder.

Reply:
fixed 7a4727322d3d7c110dd555d7f68c22d75babc643